File: Day18/Part2/merge_k_sorted_lists.py
# LC 23 — Merge k Sorted Lists
# Min-heap approach
from typing import List, Optional
import heapq

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    def __repr__(self):
        return f"{self.val}->{self.next!r}"

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        pq = []
        seq = 0
        for node in lists:
            if node:
                heapq.heappush(pq, (node.val, seq, node))
                seq += 1
        dummy = ListNode()
        cur = dummy
        while pq:
            _, _, node = heapq.heappop(pq)
            cur.next = node
            cur = cur.next
            if node.next:
                heapq.heappush(pq, (node.next.val, seq, node.next))
                seq += 1
        return dummy.next

if __name__ == "__main__":
    # Example: [[1,4,5],[1,3,4],[2,6]] -> [1,1,2,3,4,4,5,6]
    def build(lst):
        head = cur = ListNode(0)
        for v in lst:
            cur.next = ListNode(v)
            cur = cur.next
        return head.next
    lists = [build([1,4,5]), build([1,3,4]), build([2,6])]
    ans = Solution().mergeKLists(lists)
    out = []
    while ans:
        out.append(ans.val)
        ans = ans.next
    print(out)  # [1,1,2,3,4,4,5,6]

File: learnLeetCode_remote/Day18/Part2/merge_k_sorted_lists.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 23 — Merge k Sorted Lists (min-heap)

struct ListNode{
    int val; ListNode* next;
    ListNode(int v=0, ListNode* n=nullptr): val(v), next(n) {}
};

struct Cmp{
    bool operator()(ListNode* a, ListNode* b) const{
        return a->val > b->val; // min-heap
    }
};

ListNode* mergeKLists(vector<ListNode*>& lists){
    priority_queue<ListNode*, vector<ListNode*>, Cmp> pq;
    for (auto node: lists) if (node) pq.push(node);
    ListNode dummy(0); ListNode* cur = &dummy;
    while(!pq.empty()){
        ListNode* node = pq.top(); pq.pop();
        cur->next = node; cur = cur->next;
        if (node->next) pq.push(node->next);
    }
    return dummy.next;
}

// helpers for local test
ListNode* build(const vector<int>& a){
    ListNode* head = new ListNode(0); ListNode* cur=head;
    for(int v: a){ cur->next = new ListNode(v); cur=cur->next; }
    ListNode* res=head->next; delete head; return res;
}

int main(){
    vector<ListNode*> lists = { build({1,4,5}), build({1,3,4}), build({2,6}) };
    ListNode* ans = mergeKLists(lists);
    vector<int> out; while(ans){ out.push_back(ans->val); ans=ans->next; }
    for(size_t i=0;i<out.size();++i){ if(i) cout<<","; cout<<out[i]; }
    cout<<"\n"; // 1,1,2,3,4,4,5,6
    return 0;
}
